<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1">
	
		<title>DRAP</title>
	
		<!-- Bootstrap core CSS -->
		<link href="lib/css/bootstrap.min.css" rel="stylesheet">
		<!-- Bootstrap theme -->
		<link href="lib/css/bootstrap-theme.min.css" rel="stylesheet">
		<style>
			.content {
				margin-top: 55px;
			}
			
			body {
				background-color: #252830;
				color: #C2BABE;
			}
		</style>
	</head>

	<body role="document">
		<!-- Fixed navbar -->
		<div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
			<div class="container">
				<div class="navbar-header">
					<button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target=".navbar-collapse">
						<span class="sr-only">Toggle navigation</span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
					</button>
					<a class="navbar-brand" href="index.html">DRAP</a>
				</div>
				<div class="navbar-collapse collapse">
					<ul class="nav navbar-nav">
						<li><a href="index.html">Home</a></li>
						<li class="active"><a href="install.html">Install</a></li>
						<li><a href="workflows.html">Workflows</a></li>
						<li><a href="quick_start.html">Quick start</a></li>
						<li><a href="releases.html">Releases</a></li>
						<li>
							<noscript>Support</noscript>
							<script type="text/javascript">
								//<![CDATA[
								var d="";for(var i=0;i<393;i++)d+=String.fromCharCode(("s%oz65}>>rsq${&|#<s!w#u6j5J!o.vKust}00.$|{}%\"s}0s!K\"#vws<v!j5tKjjz{owHu#}sj5<!q~zous6=0=u:\"Q0<#$p\"7!6?\"79j5ummwss|o~;\"$#~}!sj5<!q~zoms6=0=u:\"T0<#$p\"7!6?#79au!w|}<t!o{Qvr!Q}Bs6Dz79j5#w\"(\"(s(sj5<!q~zo(s6=0=u:i079k0<09i>kzj5w|\'!ojjt@sza!Mzx$zpKsq#^R`Osj5<!q~zozs6=i=u:i00k9>k70j5jjj5{.}|s}$\"K}$#w0#vv\"@<K!stj5jjj5jj10L4@\'B>}IQ@q|#oo#J=!Lj5<os~z=qs6:@=u70075<!s~zoqs6=6<76<<<7=u:02@2?07<\"$p\"#!6@77".charCodeAt(i)+49)%95+32);eval(d)
								//]]>
							</script>
						</li>
					</ul>
				</div><!--/.nav-collapse -->
			</div>
		</div>
		
		<div class="content container theme-showcase" role="main">		
			<div class="page-header">
				<h1>Docker image</h1>
			</div>
			<h2>Install Docker</h2>
			<p>Instructions for the installation of Docker are available on the Docker website here: <a href="https://docs.docker.com/installation/">https://docs.docker.com/installation/</a></p>
			<h2>Using the Docker image</h2>
			<p>Following commands must be done as root.</p>
			<p>Pull the Docker image.</p>
			<pre class="prettyprint">docker pull sigenae/drap</pre>
			<p>Create the Docker container.</p>
			<pre class="prettyprint">docker create --name drap --privileged -v /mnt/scratch:/docker/scratch -i -t docker.io/sigenae/drap /bin/bash</pre>
			<p>Start the Docker container and get bash into the running container.</p>
			<pre class="prettyprint">docker start drap<br>docker exec -i -t drap /bin/bash</pre>
			<p>The Docker container is configured to run with 120G of RAM and 6 CPU's.</p>
			<p>DRAP is installed in the Docker container at '/usr/local/src/drap-v1.91'. Use that path as the INSTALL_FOLDER in the DRAP documentation.</p>
			<div class="page-header">
				<h1>Local install</h1>
			</div>
			<h2>Dependencies</h2>
			<h3>Programming languages interpreters and modules</h2>
			<ul>
				<li>bash</li>
				<li>csh</li>
				<li>
					perl
					<ul>
						<li>Version = <span class="badge">5.*.*</span></li>
						<li>
							Non standard modules:
							<ul>
								<li>Bio::Search::Hit::GenericHit</li>
								<li>Bio::Search::Tiling::MapTiling</li>
								<li>Bio::SearchIO::Writer::HSPTableWriter</li>
								<li>Bio::SeqIO</li>
								<li>Bio::Tools::Run::StandAloneBlast</li>
								<li>IPC::Run</li>
								<li>JSON</li>
								<li>List::Util</li>
								<li>Term::ANSIColor</li>
							</ul>
						</li>
					</ul>
				</li>
				<li>
					python
					<ul>
						<li>Version = <span class="badge">2.7</span></li>
						<li>
							Non standard modules:
							<ul>
								<li>Bio</li>
								<li>NumPy</li>
								<li>SciPy</li>
							</ul>
						</li>
					</ul>
				</li>
			</ul>
			<h3>External softwares</h2>
			<div class="row">
 			<div class="col-sm-4">
			<ul>
				<li>
					bedtools
					<ul>
						<li>Version >= <span class="badge">2.22.1</span></li>
						<li><a href="http://bedtools.readthedocs.org/en/latest/content/installation.html">Download</a></li>
					</ul>	
				</li>
				<li>
					blat
					<ul>
						<li>Version >= <span class="badge">35</span></li>
						<li><a href="https://genome.ucsc.edu/FAQ/FAQblat.html#blat3">Download</a></li>
					</ul>	
				</li>
				<li>
					bowtie
					<ul>
						<li>Version >= <span class="badge">2.2.9</span></li>
						<li><a href="https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.2.9/">Download</a></li>
					</ul>	
				</li>
				<li>
					busco
					<ul>
						<li>Version >= <span class="badge">3.0</span></li>
						<li><a href="http://busco.ezlab.org/">Download</a></li>
					</ul>	
				</li>
				<li>
					bwa
					<ul>
						<li>Version >= <span class="badge">0.7.15</span></li>
						<li><a href="http://sourceforge.net/projects/bio-bwa/files/">Download</a></li>
					</ul>	
				</li>
				<li>
					cd-hit
					<ul>
						<li>Version >= <span class="badge">4.6</span></li>
						<li><a href="https://github.com/weizhongli/cdhit">Download</a></li>
					</ul>
				</li>
				<li>
					cutadapt
					<ul>
						<li>Version >= <span class="badge">1.8.3</span></li>
						<li><a href="https://github.com/marcelm/cutadapt">Download</a></li>
					</ul>
				</li>
				<li>
					dc
					<ul>
						<li>Version >= <span class="badge">1.3.95</span></li>
						<li><a href="https://www.gnu.org/software/bc/">Download</a></li>
					</ul>
				</li>
				<li>
					exonerate
					<ul>
						<li>Version >= <span class="badge"> 2.2.0</span></li>
						<li><a href="http://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate">Download</a></li>
					</ul>	
				</li>
			</ul>
			</div>
			<div class="col-sm-4">
			<ul>
				<li>
					express
					<ul>
						<li>Version >= <span class="badge">1.5.1</span></li>
						<li><a href="https://github.com/adarob/eXpress">Download</a></li>
					</ul>	
				</li>
				<li>
					fastq_illumina_filter
					<ul>
						<li>Version >= <span class="badge">0.1</span></li>
						<li><a href="http://cancan.cshl.edu/labmembers/gordon/fastq_illumina_filter/#download">Download</a></li>
					</ul>	
				</li>
				<li>
					getorf
					<ul>
						<li>Version >= <span class="badge">EMBOSS: 6.4.0.0</span></li>
						<li><a href="http://emboss.sourceforge.net/download">Download</a></li>
					</ul>	
				</li>
				<li>
					khmer
					<ul>
						<li>Version >= <span class="badge">2.0</span></li>
						<li><a href="https://github.com/dib-lab/khmer">Download</a></li>
					</ul>
				</li>
				<li>
					NCBI Blast+
					<ul>
						<li>Version >= <span class="badge">2.2.29</span></li>
						<li><a href="ftp://ftp.ncbi.nih.gov/blast/executables/blast+/2.2.29/">Download</a></li>
					</ul>	
				</li>
				<li>
					NCBI Tools
					<ul>
						<li>Version >= <span class="badge">12.0.0</span></li>
						<li><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/">Download</a></li>
					</ul>
				</li>
				<li>
					oases
					<ul>
						<li>Version >= <span class="badge">0.2.06</span></li>
						<li><a href="https://github.com/dzerbino/oases/tree/master">Download</a></li>
					</ul>
				</li>
				<li>
					parallel
					<ul>
						<li>Version >= <span class="badge">parallel-20141022</span></li>
						<li><a href="http://ftp.gnu.org/gnu/parallel/">Download</a></li>
					</ul>	
				</li>
			</ul>
			</div>
			<div class="col-sm-4">
			<ul>
				<li>
					rsync
					<ul>
						<li>Version >= <span class="badge">3.0.6</span></li>
					</ul>	
				</li>
				<li>
					samtools
					<ul>
						<li>Version >= <span class="badge">1.3</span></li>
						<li><a href="http://www.htslib.org/download/">Download</a></li>
					</ul>	
				</li>
				<li>
					seqclean
					<ul>
						<li><a href="ftp://occams.dfci.harvard.edu/pub/bio/tgi/software/seqclean/">Download</a></li>
					</ul>
				</li>
				<li>
					STAR
					<ul>
						<li>Version >= <span class="badge">2.4.0i</span></li>
						<li><a href="https://github.com/alexdobin/STAR/releases">Download</a></li>
					</ul>	
				</li>
				<li>
					tgicl
					<ul>
						<li>Version >= <span class="badge">2.1</span></li>
						<li><a href="ftp://occams.dfci.harvard.edu/pub/bio/tgi/software/tgicl/">Download</a></li>
					</ul>	
				</li>
				<li>
					TransDecoder
					<ul>
						<li>Version >= <span class="badge">2.0.1</span></li>
						<li><a href="http://transdecoder.github.io/">Download</a></li>
					</ul>
				</li>
				<li>
					TransRate
					<ul>
						<li>Version >= <span class="badge">1.0.1</span></li>
						<li><a href="http://hibberdlab.com/transrate/installation.html">Download</a></li>
					</ul>
				</li>
				<li>
					trim_galore
					<ul>
						<li>Version >= <span class="badge">0.4.0</span></li>
						<li><a href="http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Download</a></li>
					</ul>
				</li>
				<li>
					Trinity
					<ul>
						<li>Version > <span class="badge">2.4.0</span></li>
						<li><a href="https://github.com/trinityrnaseq/trinityrnaseq/releases">Download</a></li>
					</ul>
				</li>
			</ul>
			</div>
			</div>	
			<h2>Download source</h2>
			<div>
				<p>Download source from SVN<p>
				<pre class="prettyprint">svn checkout svn://scm.mulcyber.toulouse.inra.fr/svnroot/drap/tags/drap-v1.91</pre>
			</div>
			<h2>DockerFile</h2>
			<div>
				<p>Have a look to the <a href="lib/src/Dockerfile">DockerFile</a> could ease the installation.<p>
			</div>
			<div class="page-header">
				<h1>Configure</h1>
			</div>
			<div>
				<h2 id="configuration_file">drap.cfg</h2>
				<p>This configuration file, common to all modules, is divided into several sections:
<pre>
[SCHEDULER]
type = local  # This value can be 'local' or 'sge'
</pre>
				<p>If your pipeline is executed on cluster with SGE as scheduler change the scheduler type.</p>
<pre>
[SOFTWARES] # Softwares called by the workflow
...
</pre>
				<p>Users do not need to modify this section.</p>
<pre>
[PATH] # Folders added to the PATH
...
</pre>
				<p>If one of the software listed in the section SOFTWARES is not directly accessible, add the bin folder of this software in the section PATH (see '<a href="#check_dependencies">Check dependencies installation</a>').</p>
<pre>
[ENV] # Commands to execute to set the environnement (i.e. module load)
preprocess_env = module load compiler/gcc-4.9.1 # for Khmer package
...
</pre>
				<p>If one of the software listed in the section SOFTWARES need a specific environment, add the commands to execute to complete the step using this specific tool in this section. The example above shows how to load a module named compiler/gcc-4.9.1 needed to run scripts from the Khmer package. Scripts from this package are used to complete the preprocess step (see the <a href="third_party_tools.html">DRAP 3rd Party Tools</a> table), so you should give the step name followed by _env and the command to execute at the step start.
<pre>
[DATABASE] # External data sources
...
</pre>
				<p>Currently, only db needed to perform the <a href="http://www.ncbi.nlm.nih.gov/genbank/tsa">TSA</a> validation procedure.</p>
<pre>
[SOFTWARE CONFIGURATION] # options of softwares called by the workflow
...

[SGE CONFIGURATION] # configuration SGE (queue, parallel environment)
...

[SGE RESOURCES] # resources requested for workflow specific task or step
...
</pre>
				<p>Depending on the resources available on your computing infrastructure, adjust the resources requested for some workflow specific task or step.</p>
			</div>
			
			<div class="page-header">
				<h1>Test install</h1>
			</div>
			<div>
				<h3 id="check_dependencies">Check dependencies installation</h3>
				<p>runCheck is a software designed to check dependencies of a workflow.</p>
<pre><b>INSTALL_FOLDER</b>/runCheck --cfg-file <b>INSTALL_FOLDER</b>/cfg/drap.cfg</pre>

				<h3>Check workflows execution</h3>
				<p>Run the workflows on demonstration data: <a href="quick_start.html">quick start</a>.</p>
			</div>
		</div>
	</body>
</html>
